README.md
# MongoFB

## Table of Contents

 - [General Information](#general-information)
  - [What is MongoFB](#what-is-mongofb)
  - [Diagram](#diagram)
 - [Usage](#usage)
  - [Server Configuration](#index-configuration)
  - [Server API](#index-api)
  - [Server Hooks](#index-hooks)
  - [Authentication](#authentication)
  - [Client SDK](#mongoFbClient-sdk)
 - [Examples](#examples)
  - [Server](#index)
  - [Javascript Client](#javascript-mongoFbClient)
  - [iOS Client](#ios-mongoFbClient)
  - [Android Client](#android-mongoFbClient)

## General Information

### What is MongoFB

MongoFB is a combination of MongoDB and Firebase. You can run your MongoDB
anywhere you like, and sign up for your own Firebase account.  MongoFB
provides an API so you can query, index, and aggregate your data (all the
things you wish your firebase could do).  MongoFB uses Firebase as its
master source of data, so you get built in Security Rules, Authentication,
and can listen for updates on any document of any collection, or any field
of any document (all the things you wish your mongodb would do).

 - Firebase
   - Security/Authentication
   - WebSockets
 - MongoDB
   - querying
   - indexing
   - aggregation

### Diagram
![Diagram](http://media.lessthan3.com/wp-content/uploads/2013/07/mongofb.png)

## Usage

### Sever Configuration

MongoFB can be included in your express or zappajs index, using middleware.
```
mongofb = require 'mongofb'

app.use mongofb {
  root: '/api/1'      # the root url to host mongofb on
  cache:
    max: 100          # the max number of results to store in a local LRU cache
    maxAge: 1000*60*5 # the max age of any result in the LRU cache
  firebase:
    url: ''           # the url of your firebase
    secret: ''        # your firebase secret - only needed if your firebase has security rules
  mongodb:
    db: 'test'        # the mongodb to connect to
    host: 'localhost' # the host of your mongodb
    pass: ''          # the password to connect with
    port: 27017       # the port to connect to
    user: 'admin'     # the user to connect with
    options: {}       # other connection options [ref](#https://github.com/mongodb/node-mongodb-native/blob/master/docs/articles/MongoClient.md#basic-parts-of-the-url)
}
```


### Server API

/API-ROOT/mongofb.js
```
Serves the javascript mongoFbClient
```

/API-ROOT/Firebase
```
Let's the mongoFbClient look up the public url of your Firebase
```

/API-ROOT/ObjectId
```
The mongoFbClient calls here to get a new ObjectID before writing to Firebase
```

/API-ROOT/sync/:collection/:id
```
After an insert, update, or remove, a mongoFbClient will tell the index it needs to
update data in Firebase. The index will then pull the most up-to-date data
directly from Firebase and write it to MongoDB for querying.
```

/API-ROOT/:collection/find
```
Perform a db.collection.find on your MongoDB. Pass your query as query
parameters to this endpoint. The result is returned as an array.

special options
 - limit: limits the number of results in the response
```

/API-ROOT/:collection/findOne
```
Perform a db.collection.findOne on your MongoDB. Pass your query as query
parameters to this endpoint. The result is returned as an object
```

/API-ROOT/:collection/:id*
```
Perform a db.collection.findOne by {_id: ObjectID()} on your MongoDB. Pass your
query as query parameters to this endpoint. The result is returned as an object.

This endpoint functions more like a standard resource url as no query parameters
are used.  This method also lets you query for specific fields of a document.

example: /API-ROOT/posts/510b56c221168da296f27bd5/author/name

The above might be a posts collection for my blog. With this I could directly
access the author's name of post 510b56c221168da296f27bd5.

The corresponding Firebase URL for that data would be
https://my-firebase.firebaseio.com/posts/510b56c221168da296f27bd5/author/name
```

### Server Hooks
Sometimes you may want to modify the response from your api, or set default
values for parameters, or do something special if the user is authenticated.
This is all possible with MongoFB Hooks.

You can define your hooks in your index configuration. The current hooks
available are...

new_query = collection.before.find(query)

new_doc = collection.after.find(doc)


Example Usage
```
app.use mongofb {
  firebase: config.firebase
  mongofb: config.mongodb
  root: '/api/1'
  hooks:
    users:
      after:
        find: (doc) ->
          # hide private user information to other users
          return doc if @user?.auth?.id == doc.id
          {_id: doc._id, public: doc.public}
    posts:
      before:
        find: (query) ->
          # an author changed their name
          if query.author?.name == 'joe'
            query.author.name = 'joey'

          # if we search for football or baseball, also search all sports
          if query.tag in ['football', 'baseball']
            query.tag = [query.tag, 'sports']

          # force a small limit
          query.limit = 10
}
```

### Authentication
authenticate any request by passing a token query parameter with the
value being the users' firebase token.

The @user can then be referenced in your hooks

### Client SDK
How to use the Javascript SDK

Classes
```
mongofb.Database
mongofb.Collection
mongofb.Document
mongofb.DocumentRef
```

mongofb.Database
```
# This is the equivalent to a MongoDB Database

# Connect to our MongoFB index
db = new mongofb.Database 'http://localhost:3000/API-ROOT'

# Get a collection
posts = db.collection 'posts'
posts = db.get 'posts'

# Get a document directly
post = db.collection('posts').get('510b56c221168da296f27bd5')
post = db.get('posts/510b56c221168da296f27bd5')
post = db.get('posts.510b56c221168da296f27bd5')

# Get a field from a document directly
name = db.get('posts/510b56c221168da296f27bd5/author/name')
name = db.get('posts.510b56c221168da296f27bd5.author.name')
```

mongofb.Collection
```
# This is the equivalent to a MongoDB Collection

# Get a collection
users = db.collection 'users'
users = db.get 'users'

# Insert a document (this method must be asynchronous)
users.insert {foo: 'bar'}, (err, user) ->
  throw err if err
  console.log user.val()

# Run a find query (synchronous)
docs = users.find {foo: 'bar'}

# Run a find query (asynchronous)
users.find {foo: 'bar'}, (err, docs) ->
  throw err if err
  console.log docs

# Run a findById (synchronous)
user = users.findById '510b56c221168da296f27bd5'

# Run a findById (asynchronous)
users.findById '510b56c221168da296f27bd5', (err, user) ->
  throw err if err
  console.log user

# Run a findOne (synchronous)
user = users.findOne {foo: 'bar'}

# Run a findOne (asynchonous)
users.findOne {foo: 'bar'}, (err, user) ->
  throw err if err
  console.log user

# Remove a document (this method must be asynchronous)
# only allowed to remove by id
users.remove '510b56c221168da296f27bd5', (err) ->
  throw err if err
```

mongofb.Document
```
# This is the equivalent of a MongoDB Document
post = posts.findById '510b56c221168da296f27bd5'

# Update a field in a Document
post.get('author.name').set('new author')

# update an entire Document
post.set {author: {name: 'the author'}, content: 'long post'}

# get json for a document
post.val()

# listeners
post.on 'update', (val) ->
  # called when this document is updated

post.on 'value', (val) ->
  # called immediately, and when the document is updated

post.on 'remove', (val) ->
  # called when the post is removed from the database
```

mongofb.DocumentRef
```
# A DocumentRef is a reference to a field of a Document

# Get a ref
ref = post.get('author.name')

# add listeners
ref.on 'update', (val) ->
ref.on 'value', (val) ->

# remove listeners
ref.off 'update'
ref.off 'value'
ref.off()

# get the parent ref
ref.parent()

# change the value of this property
ref.set('new author')

# get the json value for this ref
ref.val()


```

## Examples

### Server
```
express = require 'express'
mongofb = require '../lib/index'

app = express()
app.use mongofb {
  firebase:
    url: 'https://vn42xl9zsez.firebaseio-demo.com/'
  mongodb:
    host: 'localhost'
    port: 27017
  root: '/api/v2'
}
app.get '/', (req, res) ->
  res.send """
  <html>
    <body>
      <script type='text/javascript' src='http://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.2/jquery.min.js'></script>
      <script type='text/javascript' src='https://cdn.firebase.com/v0/firebase.js'></script>
      <script type='text/javascript' src='/api/v2/mongofb.js'></script>
    </body>
  </html>
  """

app.listen 3000
console.log "listening: 3000"
```


### Javascript Client
```
window.db = new mongofb.Database 'http://localhost:3000/api/v2'
window.cookies = db.collection 'cookies'

cookies.insert {type: 'chocolate'}, (err, cookie) ->
  throw err if err
  window.cookie = cookie

  cookie.on 'update', (val) ->
    console.log 'cookie updated to', val

  ref = cookie.get 'type'
  ref.on 'update', (val) ->
    console.log 'cookie.type updated to', val

  ref.set 'peanut butter'
```

### iOS Client
```
Coming Soon!
```

### Android Client
```
Coming Soon!
```

### dependencies required for testing
```
  "devDependencies": {
    "async": "2.0.1",
    "body-parser": "1.6.x",
    "coffee-script": "1.10.0",
    "teacup": "2.0.0"
  },
```




CHANGELOG.md
**1.0.0** (2016-08-12)
 - fork to dobi-mongofb

**0.9.15** (2015-05-13)

 - leave token in authenticated requests

**0.9.14** (2015-05-08)

 - do not allow $where queries

**0.9.13** (2015-02-17)

 - fix error response for invalid ObjectIDs

**v0.9.12** (2015-01-17)

 - change 500 errors to 400s

**v0.9.11** (2014-12-04)

 - allow multiple DocumentRefs to listen to the same property

**v0.9.10** (2014-10-28)

 - emit updates after the document.data has been updated
 - default @data to null because firebase data is never undefined
 - fix isEqual when a and b are both null
 - fix val() when @data is null, typeof(null) is 'object'
 - add global counter to DocumentRefs to help debug

**v0.9.9** (2014-10-21)

 - fix isEqual for when refs do not reference objects

**v0.9.7** (2014-08-17)

 - set required node version in package.json
 - add travis test configuration

**v0.9.6** (2014-08-05)

 - also check POST requests for a 'token' to authenticate the user

**v0.9.5** (2014-07-13)

 -  add minified mongoFbClient js route

**v0.9.4** (2014-07-07)

 - change 'callback' to 'handler' for events to avoid confusion
 - better cache bust for inserts

**v0.9.3** (2014-05-25)

 - allow Collection.find()
 - allow Collection.find(callback)

**v0.9.2** (2014-05-25)

 - fix PseudoCollection constructor

**v0.9.1** (2014-05-25)

 - add input validation to Collection.remove
 - add DocumentRef.remove

**v0.9.0** (2014-05-25)

 - add PseudoCollection
 - change Collection.remove to Collection.removeById to avoid confusion

**v0.8.2** (2014-05-25)

 - only check created and last_modified fields on document level refs

**v0.8.1** (2014-05-24)

 - fix updateData bug for when @data.created is undefined

**v0.8.0** (2014-05-23)

 - Add auto created=Date.now() config option
 - ignore 'created' and 'last_modified' fields for triggering updates

**v0.7.8** (2014-05-23)

 - Date.now() should be enough

**v0.7.7** (2014-05-23)

 - fix caching when running index-side

**v0.7.6** (2014-05-23)

 - add npm-debug.log to gitignore
 - handle 500 responses as errors from index-side usage fetches

**v0.7.5** (2014-05-22)

 - body = JSON.parse body bug

**v0.7.4** (2014-05-22)

 - need to keep next(null, null) for 404 case from fetch

**v0.7.3** (2014-05-22)

 - 0.7.2, but let 404s return as null instead of error

**v0.7.2** (2014-05-22)

 - better handling of bad request responses

**v0.7.1** (2014-05-15)

 - better error messaging

**v0.7.0** (2014-05-01)

 - update asset-wrap dependency

**v0.6.4** (2014-03-08)

 - add last_modified feature

**v0.6.3** (2014-03-08)

 - add fields to findById query

**v0.6.2** (2014-03-08)

 - add built-in hooks for common operations

**v0.6.1** (2014-03-06)

 - rollback firebase data if mongodb sync fails

**v0.6.0** (2014-02-03)

 - update asset-wrap to 0.6.x
 - switch to jwt parsing library for token check
 - watch out for invalid query parameters
 - update other dependencies

**v0.5.8** (2014-02-03)

 - update asset-wrap to 0.5.x

**v0.5.7** (2014-01-20)

 - allow fields for non json http requests

**v0.5.6** (2014-01-17)

 - default values of query field should be null

**v0.5.5** (2014-01-17)

 - send 404 when findOne finds nothing

**v0.5.4** (2014-01-16)

 - add a new find query object for token and nocache
 - multiple ways to use find and findOne arguments

**v0.5.3** (2014-01-15)

 - add fields and options to javascript SDK
 - only allow writes to queried fields

**v0.5.2** (2014-01-13)

 - fix example for v0.4.x+
 - fix duplicate update bug
 - add .gitignore

**v0.5.1** (2014-01-13)

 - add ObjectID to exports
 - allow multiple arguments to hooks
 - hooks edit object instead of returning new
 - no max limit (this should be set through a hook)
 - no default limit (this should be set through a hook)
 - refactor findOne and findById to route through find
 - allow json arguments for criteria and fields
 - add field filtering
 - add sort and skip

**v0.4.1** (2013-12-27)

 - emit object clone for events

**v0.3.14** (2013-11-22)

 - add Database.auth
 - fix index-side ObjectID fetching (response is not json)

**v0.3.13** (2013-11-22)

 - add request dependency

**v0.3.12** (2013-11-04)

 - setToken function for mongoFbClient
 - check for statusCode on index-side mongoFbClient.coffee fetch
 - check for admin property in authentication
 - add mongofb reference to incoming requests

**v0.3.11** (2013-11-02)

 - for browser-side errors, return jqXHR object instead of error string

**v0.3.10** (2013-10-30)

 - more verbose cache bust for ObjectID (cant risk collisions here)

**v0.3.9** (2013-10-30)

 - dont set firebase created timestamp on new docs (use custom or ObjectID)
 - check for Array before object when getting val()

**v0.3.8** (2013-10-30)

 - simpler way to generate ObjectIDs
 - handle exceptions for bad ObjectIDs to sync and findByID

**v0.3.7** (2013-10-29)

 - pass json to fetch (needed for index-side mongofb mongoFbClient)
 - set correct limit parameter for finds

**v0.3.6** (2013-10-26)

 - cache avoid for ObjectId and sync on inserts

**v0.3.5** (2013-10-26)

 - expose fb and db

**v0.3.4** (2013-10-23)

 - fix EventEmitter.off to remove if the callback matched

lib/server/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.server = exports.ObjectID = exports.client = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _express = _interopRequireDefault(require("express"));

var _firebase = _interopRequireDefault(require("firebase"));

var _firebaseTokenGenerator = _interopRequireDefault(require("firebase-token-generator"));

var _fs = _interopRequireDefault(require("fs"));

var _jwtSimple = _interopRequireDefault(require("jwt-simple"));

var _lruCache = _interopRequireDefault(require("lru-cache"));

var _deepmerge = _interopRequireDefault(require("deepmerge"));

var _mongodb = _interopRequireDefault(require("mongodb"));

var mongoFbClient = _interopRequireDefault(require("../client"));

var _dirname = require("./dirname");

/* eslint-disable no-underscore-dangle */
// dependencies
var client = mongoFbClient;
exports.client = client;
var ObjectID = _mongodb.default.ObjectID; // exports

exports.ObjectID = ObjectID;

var server = function server(_cfg) {
  var cfg = _cfg; // configuration

  cfg = (0, _deepmerge.default)({
    cache: {
      enabled: true,
      max: 100,
      maxAge: 1000 * 60 * 5
    },
    firebase: {
      secret: null,
      url: 'https://shining-fire-369.firebaseio.com/'
    },
    mongodb: {
      db: 'test',
      host: 'localhost',
      options: {
        autoReconnect: true,
        keepAlive: 120,
        native_parser: false,
        poolSize: 1
      },
      pass: 'testPassword',
      port: 27017,
      user: 'testUser'
    },
    options: {
      blacklist: [],
      limit_default: 20,
      limit_max: 1000,
      set_created: true,
      set_last_modified: true,
      use_objectid: true
    },
    root: '/api'
  }, cfg); // variables

  var db = null;
  var fb = null; // connect to firebase and mongodb

  var connect = function connect(_next) {
    var next = _next;

    if (typeof next !== 'function') {
      next = function next() {};
    }

    if (db && fb) {
      return next();
    }

    var _cfg$mongodb = cfg.mongodb,
        database = _cfg$mongodb.db,
        host = _cfg$mongodb.host,
        options = _cfg$mongodb.options,
        pass = _cfg$mongodb.pass,
        port = _cfg$mongodb.port,
        user = _cfg$mongodb.user;
    var url = "mongodb://".concat(user, ":").concat(pass, "@").concat(host, ":").concat(port, "/").concat(database).replace(':@', '@');
    return _mongodb.default.MongoClient.connect(url, options, function (err, mongoClient) {
      if (err) {
        return next(err);
      }

      db = mongoClient.db(database);
      db.ObjectID = _mongodb.default.ObjectID;
      fb = new _firebase.default(cfg.firebase.url);

      if (cfg.firebase.secret) {
        var tokenGenerator = new _firebaseTokenGenerator.default(cfg.firebase.secret);
        var token = tokenGenerator.createToken({}, {
          admin: true,
          expires: Date.now() + 1000 * 60 * 60 * 24 * 30
        });
        return fb.authWithCustomToken(token, function (authErr) {
          if (authErr) {
            return next(authErr);
          }

          fb.admin_token = token;
          return next();
        });
      }

      return next();
    });
  };

  connect(); // middleware

  return function (_parentReq, _parentRes, next) {
    return connect(function (_err) {
      var parentReq = _parentReq;
      var parentRes = _parentRes;
      var err = _err;

      if (err) {
        return next(err);
      } // databases


      parentReq.db = db;
      parentReq.fb = fb;
      parentReq.mongofb = new mongoFbClient.Database({
        firebase: cfg.firebase.url,
        server: "http://".concat(parentReq.get('host')).concat(cfg.root)
      }); // helpers

      var auth = function auth(req, res, callback) {
        var token = req.query.token || req.body.token;

        if (token && cfg.firebase) {
          try {
            var payload = _jwtSimple.default.decode(token, cfg.firebase.secret);

            req.user = payload.d;
            req.admin = payload.admin;
          } catch (authErr) {
            req.token_parse_error = authErr;
          }
        }

        return typeof callback === 'function' ? callback() : undefined;
      };

      var hasPermission = function hasPermission() {
        auth(parentReq, parentRes);

        if (parentReq.admin) {
          return true;
        }

        if (Array.from(cfg.options.blacklist).includes(parentReq.params.collection)) {
          return false;
        }

        return true;
      };

      var _cache = new _lruCache.default(cfg.cache);

      var cache = function cache(fn) {
        var maxAge = cfg.cache.maxAge / 1000;

        if (parentReq.query.bust === '1') {
          maxAge = 0;
        }

        var val = 'private, max-age=0, no-cache, no-store, must-revalidate';

        if (cfg.cache.enabled && maxAge > 0) {
          val = "public, max-age=".concat(maxAge, ", must-revalidate");
        }

        parentRes.set('Cache-Control', val);
        var key = parentReq.url.replace('&bust=1', '');

        if (parentReq.query.bust === '1') {
          _cache.del(key);

          delete parentReq.query.bust;
        }

        if (cfg.cache.enabled && _cache.has(key)) {
          return parentRes.send(_cache.get(key));
        }

        delete parentReq.query._;
        return fn(function (data) {
          _cache.set(key, data);

          return parentRes.send(data);
        });
      };

      var contentType = function contentType(type) {
        return parentRes.set('Content-Type', type);
      };

      var handleError = function handleError(handleErr) {
        contentType('text/plain');
        return parentRes.status(400).send(handleErr.toString());
      };

      var hook = function hook(time, method, _args) {
        var args = _args;
        var fn;
        var collection = parentReq.params.collection;

        if (cfg.hooks && cfg.hooks[collection] && cfg.hooks[collection] && cfg.hooks[collection][time] && cfg.hooks[collection][time][method]) {
          fn = cfg.hooks[collection][time][method];

          if (Array.isArray(args)) {
            args = [args];
          }

          return fn.apply(parentReq, args);
        }

        return args;
      }; // routes


      var router = _express.default.Router(); // fix query parameters


      router.use("".concat(cfg.root, "/*"), function (req, res, callback) {
        var map = {
          false: false,
          null: null,
          true: true
        };

        var _arr = Object.keys(req.query);

        for (var _i = 0; _i < _arr.length; _i++) {
          var k = _arr[_i];
          var v = req.query[k];

          if (Object.keys(map).includes(v)) {
            req.query[k] = map[v];
          }
        }

        return callback();
      }); // mongoFbClient javascript

      router.get("".concat(cfg.root, "/mongofb.js"), function () {
        contentType('text/javascript');
        return cache(function (callback) {
          try {
            var script = _fs.default.readFileSync("".concat(_dirname.dirname, "/../../dist/client.js"), 'utf-8');

            return callback(script);
          } catch (readErr) {
            return handleError(readErr);
          }
        });
      }); // mongoFbClient javascript minified

      router.get("".concat(cfg.root, "/mongofb.min.js"), function () {
        contentType('text/javascript');
        return cache(function (callback) {
          try {
            var script = _fs.default.readFileSync("".concat(_dirname.dirname, "/../../dist/client.min.js"), 'utf-8');

            return callback(script);
          } catch (readErr) {
            return handleError(readErr);
          }
        });
      }); // firebase url

      router.get("".concat(cfg.root, "/Firebase"), function (req, res) {
        return res.send(cfg.firebase.url);
      }); // ObjectID for creating documents

      router.get("".concat(cfg.root, "/ObjectID"), function (req, res) {
        return res.send(_mongodb.default.ObjectID().toString());
      }); // sync data from firebase
      // NOTE: requires _id to be an ObjectID
      // db.collection.update
      // db.collection.insert
      // db.collection.remove
      // the format is /sync/:collection/:id and not /:collection/:sync/:id to
      // match firebase urls. the key in firebase is /:collection/:id

      var url = "".concat(cfg.root, "/sync/:collection/:id*");
      router.get(url, auth, function (_req, _res) {
        var req = _req;
        var res = _res;
        var collection = db.collection(req.params.collection); // get data

        var ref = fb.child("".concat(req.params.collection, "/").concat(req.params.id));
        return ref.once('value', function (snapshot) {
          var qry;
          var doc = snapshot.val(); // convert _id if using ObjectIDs

          if (cfg.options.use_objectid) {
            try {
              qry = {
                _id: new _mongodb.default.ObjectID(req.params.id)
              };
            } catch (error) {
              err = error;
              return handleError('Invalid ObjectID');
            }
          } // send null if collection in blacklist
          // WARNING: IF YOU ATTEMPT TO SYNC A COLLECTION ITEM NOT
          // ON YOUR FIREBASE AND CALL SYNC, IT WILL GET DESTROYED


          if (!hasPermission()) {
            return res.send(null);
          } // insert/update


          if (doc) {
            // set created
            if (cfg.options.set_created) {
              if (doc.created == null) {
                doc.created = Date.now();
              }
            } // set last modified


            if (cfg.options.set_last_modified) {
              doc.last_modified = Date.now();
            }

            doc._id = qry._id;
            var opt = {
              safe: true,
              upsert: true
            };
            return collection.update(qry, doc, opt, function (updateErr) {
              if (updateErr) {
                return handleError(updateErr);
              }

              hook('after', 'find', doc);
              return res.send(doc);
            }); // remove
          }

          return collection.remove(qry, function (removeErr) {
            if (removeErr) {
              return handleError(removeErr);
            }

            return res.send(null);
          });
        });
      }); // db.collection.find

      url = "".concat(cfg.root, "/:collection/find");
      router.get(url, auth, function (_req, _res) {
        return cache(function (callback) {
          var req = _req;
          var res = _res; // special options (mainly for use by findByID and findOne)

          var __single = req.query.__single || false;

          var __field = null;

          if (req.query.__field) {
            __field = decodeURIComponent(req.query.__field).replace(/\//g, '.');
          }

          delete req.query.__single;
          delete req.query.__field; // defaults

          var criteria = {};
          var fields = {};
          var options = {}; // use JSON encoded parameters

          if (req.query.criteria || req.query.options) {
            if (req.query.criteria) {
              try {
                criteria = JSON.parse(req.query.criteria);
              } catch (error) {
                err = error;
                return handleError('invalid criteria');
              }
            }

            if (req.query.fields) {
              try {
                fields = JSON.parse(req.query.fields);
              } catch (error1) {
                err = error1;
                return handleError('invalid fields');
              }
            }

            if (req.query.options) {
              try {
                options = JSON.parse(req.query.options);
              } catch (error2) {
                err = error2;
                return handleError('invalid options');
              }
            } // simple http queries

          } else {
            if (req.query.fields) {
              var _iteratorNormalCompletion = true;
              var _didIteratorError = false;
              var _iteratorError = undefined;

              try {
                for (var _iterator = req.query.fields.split(',')[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                  var field = _step.value;
                  fields[field] = 1;
                }
              } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
              } finally {
                try {
                  if (!_iteratorNormalCompletion && _iterator.return != null) {
                    _iterator.return();
                  }
                } finally {
                  if (_didIteratorError) {
                    throw _iteratorError;
                  }
                }
              }

              delete req.query.fields;
            }

            if (req.query.limit) {
              options.limit = req.query.limit;
              delete req.query.limit;
            }

            if (req.query.skip) {
              options.skip = req.query.skip;
              delete req.query.skip;
            }

            if (req.query.sort) {
              var _req$query$sort$split = req.query.sort.split(','),
                  _req$query$sort$split2 = (0, _slicedToArray2.default)(_req$query$sort$split, 2),
                  sortField = _req$query$sort$split2[0],
                  sortDir = _req$query$sort$split2[1];

              options.sort = [[sortField, sortDir || 'asc']];
              delete req.query.sort;
            }

            criteria = req.query;
          }

          if (__single) {
            options.limit = 1;
          } // built-in hooks


          if (cfg.options.use_objectid) {
            try {
              if (criteria._id) {
                if (typeof criteria._id === 'string') {
                  criteria._id = new _mongodb.default.ObjectID(criteria._id);
                } else if (criteria._id.$in) {
                  var ids = criteria._id.$in;
                  criteria._id.$in = Array.from(ids).map(function (id) {
                    return new _mongodb.default.ObjectID(id);
                  });
                }
              }
            } catch (error3) {
              err = error3;
              return handleError('Invalid ObjectID');
            }
          }

          if (cfg.options.limit_default) {
            if (options.limit == null) {
              options.limit = cfg.options.limit_default;
            }
          }

          if (cfg.options.limit_max) {
            options.limit = Math.min(options.limit, cfg.options.limit_max);
          } // don't allow $where clauses in the criteria


          if (criteria.$where) {
            return res.send(403, 'use of the $where operator is not allowed');
          } // dont make a DB call for blacklist
          // as mongo errors can be revealing


          if (!hasPermission()) {
            if (__single) {
              return res.sendStatus(404);
            }

            return callback([]);
          } // hooks


          hook('before', 'find', [criteria, fields, options]); // run query

          var collection = db.collection(req.params.collection);

          if (Object.keys(fields).length > 0) {
            options.projection = fields;
          }

          return collection.find(criteria, options).toArray(function (findErr, _docs) {
            var docs = _docs;
            var doc;

            if (findErr) {
              return handleError(findErr);
            }

            var _iteratorNormalCompletion2 = true;
            var _didIteratorError2 = false;
            var _iteratorError2 = undefined;

            try {
              for (var _iterator2 = docs[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                doc = _step2.value;
                hook('after', 'find', doc);
              }
            } catch (err) {
              _didIteratorError2 = true;
              _iteratorError2 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
                  _iterator2.return();
                }
              } finally {
                if (_didIteratorError2) {
                  throw _iteratorError2;
                }
              }
            }

            if (__field) {
              var fn = function fn(_o) {
                var o = _o;
                var _iteratorNormalCompletion3 = true;
                var _didIteratorError3 = false;
                var _iteratorError3 = undefined;

                try {
                  for (var _iterator3 = __field.split('.')[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
                    var key = _step3.value;
                    o = o != null ? o[key] : undefined;
                  }
                } catch (err) {
                  _didIteratorError3 = true;
                  _iteratorError3 = err;
                } finally {
                  try {
                    if (!_iteratorNormalCompletion3 && _iterator3.return != null) {
                      _iterator3.return();
                    }
                  } finally {
                    if (_didIteratorError3) {
                      throw _iteratorError3;
                    }
                  }
                }

                return o;
              };

              docs = function () {
                var result = [];
                var _iteratorNormalCompletion4 = true;
                var _didIteratorError4 = false;
                var _iteratorError4 = undefined;

                try {
                  for (var _iterator4 = docs[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
                    doc = _step4.value;
                    result.push(fn(doc));
                  }
                } catch (err) {
                  _didIteratorError4 = true;
                  _iteratorError4 = err;
                } finally {
                  try {
                    if (!_iteratorNormalCompletion4 && _iterator4.return != null) {
                      _iterator4.return();
                    }
                  } finally {
                    if (_didIteratorError4) {
                      throw _iteratorError4;
                    }
                  }
                }

                return result;
              }();
            }

            if (__single) {
              if (docs.length === 0) {
                return res.sendStatus(404);
              }

              var _docs2 = docs;

              var _docs3 = (0, _slicedToArray2.default)(_docs2, 1);

              docs = _docs3[0];
            }

            return callback(docs);
          });
        });
      }); // db.collection.findOne

      url = "".concat(cfg.root, "/:collection/findOne");
      router.get(url, auth, function (_req, _res, getNext) {
        var req = _req;
        var res = _res;
        req.url = "".concat(cfg.root, "/").concat(req.params.collection, "/find");
        req.query.__single = true;
        return router.handle(req, res, getNext);
      }); // db.collection.findById

      url = "".concat(cfg.root, "/:collection/:id*");
      router.get(url, auth, function (_req, _res, getCallback) {
        var req = _req;
        var res = _res;
        req.url = "".concat(cfg.root, "/").concat(req.params.collection, "/find");
        req.query.criteria = JSON.stringify({
          _id: req.params.id
        });
        req.query.__single = true;

        if (req.params[1]) {
          var _req$params = (0, _slicedToArray2.default)(req.params, 2),
              field = _req$params[1];

          req.query.__field = field;
        }

        return router.handle(req, res, getCallback);
      }); // execute routes

      return router.handle(parentReq, parentRes, next);
    });
  };
};

exports.server = server;